packages/cli/src/modules/insights/insights.compaction.service.ts

```diff
@@ -0,0 +1,97 @@
+import { Service } from '@n8n/di';
+
+import { InsightsByPeriodRepository } from './database/repositories/insights-by-period.repository';
+import { InsightsRawRepository } from './database/repositories/insights-raw.repository';
+import { InsightsConfig } from './insights.config';
+
+@Service()
+export class InsightsCompactionService {
+	private compactInsightsTimer: NodeJS.Timer | undefined;
+
+	constructor(
+		private readonly insightsByPeriodRepository: InsightsByPeriodRepository,
+		private readonly insightsRawRepository: InsightsRawRepository,
+		private readonly insightsConfig: InsightsConfig,
+	) {}
+
+	startCompactionTimer() {
+		this.stopCompactionTimer();
+		this.compactInsightsTimer = setInterval(
+			async () => await this.compactInsights(),
+			this.insightsConfig.compactionIntervalMinutes * 60 * 1000,
+		);
+	}
+
+	stopCompactionTimer() {
+		if (this.compactInsightsTimer !== undefined) {
+			clearInterval(this.compactInsightsTimer);
+			this.compactInsightsTimer = undefined;
+		}
+	}
+
+	async compactInsights() {
+		let numberOfCompactedRawData: number;
+
+		// Compact raw data to hourly aggregates
+		do {
+			numberOfCompactedRawData = await this.compactRawToHour();
+		} while (numberOfCompactedRawData > 0);
+
+		let numberOfCompactedHourData: number;
+
+		// Compact hourly data to daily aggregates
+		do {
+			numberOfCompactedHourData = await this.compactHourToDay();
+		} while (numberOfCompactedHourData > 0);
+
+		let numberOfCompactedDayData: number;
+		// Compact daily data to weekly aggregates
+		do {
+			numberOfCompactedDayData = await this.compactDayToWeek();
+		} while (numberOfCompactedDayData > 0);
+	}
+
+	// Compacts raw data to hourly aggregates
+	async compactRawToHour() {
```

Comment:
Nit - can we use this style of comments? `/** */`

This allows IDEs to show them on hover.